hera_cal/nucal.py

```diff
@@ -1013,3 +1032,516 @@ def project_u_model_comps_on_spec_axis(u_model_comps, spectral_filters):
+            If True, the model visibilities will be returned.
+
+        Returns:
+        -------
+        gains : dictionary
+            Dictionary mapping antenna numbers to complex gains. Keys are antenna numbers and values are complex gains.
+        model_parameters : dictionary
+            Dictionary containing the model parameters for each polarization. Keys are polarization strings and values are
+        metadata : dictionary
+            Dictionary containing metadata from the optimization. Contains dictionaries for each polarization with the number of 
+            iterations ("niter") and the loss history of the gradient descent ("loss_history").
+        
+        If return_model:
+            model : DataContainer
+                DataContainer containing the model visibilities. DataContainer is of the form {(ant1, ant2, pol): np.array([Ntimes, Nfreqs])}
+        """
+        # Compute spectral and spatial filters
+        self._compute_filters(
+            freqs=data.freqs, spectral_filter_half_width=spectral_filter_half_width, 
+            spatial_filter_half_width=spatial_filter_half_width, eigenval_cutoff=eigenval_cutoff,
+            umin=umin, umax=umax
+        )
+
+        # Assert that the optimizer is valid
+        assert optimizer_name in OPTIMIZERS, f"Optimizer must be one of {OPTIMIZERS}. Got {optimizer_name}."
+        optimizer = OPTIMIZERS[optimizer_name](learning_rate=learning_rate)
+
+        # Compute the estimates of the model components from the data
+        if spatial_estimate_only:
+            init_model_comps = fit_nucal_foreground_model(
+                data, data_wgts, self.radial_reds, self.spatial_filters, solver=linear_solver, share_fg_model=share_fg_model, 
+                return_model_comps=True, tol=linear_tol
+            )
+            init_model_comps = project_u_model_comps_on_spec_axis(init_model_comps, self.spectral_filters)
+        else:
+            init_model_comps = fit_nucal_foreground_model(
+                data, data_wgts, self.radial_reds, self.spatial_filters, solver=linear_solver, share_fg_model=share_fg_model, 
+                spectral_filters=self.spectral_filters, return_model_comps=True, tol=linear_tol
+            )
+            
+        # Compute idealized baseline vectors from antenna positions and calibration flags
+        idealized_antpos = abscal._get_idealized_antpos(cal_flags, self.antpos, data.pols())
+
+        if estimate_degeneracies:
+            model = evaluate_foreground_model(
+                self.radial_reds, init_model_comps, spatial_filters=self.spatial_filters, spectral_filters=self.spectral_filters
+            )
+            amplitude, tip_tilt = self._estimate_degeneracies(data, model, data_wgts)
+        else:
+            amplitude = {pol: np.ones((data.shape)) for pol in data.pols()}
+            tip_tilt = {
+                pol: np.zeros((idealized_antpos[list(idealized_antpos.keys())[0]].shape[0], data.shape[0], data.shape[1])) 
```

Comment:
I think this is assuming that the array is classically redundant (just 2 phase gradient degens). Is that true? Does it have to be? 

If it's much simpler to keep that assumption, then we should raise a NotImplementedError when you try to nucal an array with extra degens. If we can relax the assumption, that'd be cool... but probably not worth spending a ton of time on...

Reply:
This just takes the `antpos` from `abscal. _get_idealized_antpos` to set the number of tip-tilt degeneracies. I believe `abscal._get_idealized_antpos` can give greater than 2 dimensions, so this should be fine